materialize-s3-iceberg/transactor.go

```diff
@@ -0,0 +1,234 @@
+		t.state.BindingStates = make(map[string]*bindingState)
+	}
+
+	for _, b := range t.bindings {
+		if _, ok := t.state.BindingStates[b.stateKey]; !ok {
+			t.state.BindingStates[b.stateKey] = &bindingState{}
+		}
+	}
+
+	return nil
+
+}
+
+func (t *transactor) Load(it *m.LoadIterator, loaded func(int, json.RawMessage) error) error {
+	for it.Next() {
+		panic("driver only supports delta updates")
+	}
+	return nil
+}
+
+func (t *transactor) Store(it *m.StoreIterator) (m.StartCommitFunc, error) {
+	var ctx = it.Context()
+	var encoder *enc.ParquetEncoder
+	var group errgroup.Group
+	var states = t.state.BindingStates
+
+	startFile := func(b binding) {
+		// Start uploading a new file, either because the binding changed or because the prior file
+		// got sufficiently large.
+		r, w := io.Pipe()
+
+		key := path.Join(t.prefix, uuid.New().String()+".parquet")
+		s3path := fmt.Sprintf("s3://%s/%s", t.bucket, key)
+		states[b.stateKey].FilePaths = append(states[b.stateKey].FilePaths, s3path)
+
+		group.Go(func() error {
+			ll := log.WithFields(log.Fields{
+				"path":  s3path,
+				"table": pathToFQN(b.path),
+			})
+
+			ll.Info("started uploading file")
+			if err := t.store.PutStream(ctx, r, key); err != nil {
+				r.CloseWithError(err)
+				return fmt.Errorf("uploading file: %w", err)
+			}
+			ll.Info("finished uploading file")
+
+			return nil
+		})
+
+		encoder = enc.NewParquetEncoder(w, b.pqSchema, enc.WithParquetCompression(enc.Snappy))
+	}
+
+	finishFile := func() error {
+		if encoder == nil {
+			return nil
+		} else if err := encoder.Close(); err != nil {
+			return fmt.Errorf("closing encoder: %w", err)
+		} else if err := group.Wait(); err != nil {
+			return fmt.Errorf("group.Wait(): %w", err)
+		}
+
+		encoder = nil
+		return nil
+	}
+
+	lastBinding := -1
+	for it.Next() {
+		b := t.bindings[it.Binding]
+
+		if lastBinding != -1 && lastBinding != it.Binding {
+			if err := finishFile(); err != nil {
+				return nil, fmt.Errorf("finishFile after binding change: %w", err)
+			}
+		}
+		lastBinding = it.Binding
+
+		if encoder == nil {
+			startFile(b)
+		}
+
+		row := make([]any, 0, len(it.Key)+len(it.Values)+1)
+		row = append(row, it.Key.ToInterface()...)
+		row = append(row, it.Values.ToInterface()...)
+		if b.includeDoc {
+			row = append(row, it.RawJSON)
+		}
+
+		if err := encoder.Encode(row); err != nil {
+			return nil, fmt.Errorf("encoding row: %w", err)
+		}
+
+		if encoder.Written() > fileSizeLimit {
+			if err := finishFile(); err != nil {
+				return nil, fmt.Errorf("finishFile on file size limit: %w", err)
+			}
+		}
+	}
+	if err := it.Err(); err != nil {
+		return nil, fmt.Errorf("store iterator error: %w", err)
+	}
+
+	if err := finishFile(); err != nil {
+		return nil, fmt.Errorf("final finishFile: %w", err)
+	}
+
+	return func(ctx context.Context, runtimeCheckpoint *protocol.Checkpoint) (*pf.ConnectorState, m.OpFuture) {
+		// The driver checkpoint for this transaction will include the list of files upload for each
```

Comment:
Would you please unpack the checkpointing strategy a bit more? Perhaps as a package comment. It's going over my head at the moment (and I'm lacking mental context). A written narrative that discusses "When X then Y, else Z" would be helpful to show how these get mutated over time.